File: resources/[misc]/grids/shared/grids.lua
Grids = {
	Size = vector2(8992, 12992),
	Offset = vector2(4480, 4480),
	PlayerGridSize = 4,
	Deltas = {
		vector2(-1.0, -1.0),
		vector2(-1.0, 0.0),
		vector2(-1.0, 1.0),
		vector2(0.0, -1.0),
		vector2(0.0, 1.0),
		vector2(1.0, -1.0),
		vector2(1.0, 0.0),
		vector2(1.0, 1.0),
	},
	Sizes = {
		16,
		32,
		64,
		128,
	},
}

--[[ Functions ]]--
local function round(number)
	if number % 1.0 >= 0.5 then
		return math.ceil(number)
	else
		return math.floor(number)
	end
end

--[[ Functions: Grids ]]--
function Grids:GetWidth(size)
	return math.ceil(self.Size.x / size)
end

function Grids:GetGrid(coords, size, keepSize)
	coords = vector3(math.min(math.max(coords.x + self.Offset.x, 0), self.Size.x), math.min(math.max(coords.y + self.Offset.y, 0), self.Size.y), coords.z)
	if not keepSize then
		size = self.Sizes[size]
	end
	local width = self:GetWidth(size)
	return math.floor(coords.x / size) + math.floor(coords.y / size) * width
end

function Grids:GetCoords(grid, size, keepSize)
	if not keepSize then
		size = self.Sizes[size]
	end
	local width = self:GetWidth(size)
	return vector3(math.floor(grid % width) * size - self.Offset.x + size / 2.0, math.floor(grid / width) * size - self.Offset.y + size / 2.0, 0)
end

function Grids:GetNearbyGrids(coords, size, keepSize)
	if not size then return {} end
	if not keepSize then
		size = self.Sizes[size]
	end

	local width = self:GetWidth(size)
	if type(coords) == "vector3" then
		coords = self:GetGrid(coords, size, true)
	end

	local grid = coords
	
	return {
		grid,
		grid - 1,
		grid + 1,
		grid - width,
		grid - width - 1,
		grid - width + 1,
		grid + width,
		grid + width - 1,
		grid + width + 1
	}
end

function Grids:GetImmediateGrids(coords, size, keepSize)
	if not size then return {} end
	if not keepSize then
		size = self.Sizes[size]
	end

	if type(coords) == "vector4" then
		coords = vector3(coords.x, coords.y, coords.z)
	end

	local grid = self:GetGrid(coords, size, true)
	local cache = { [grid] = true }
	local grids = { grid }
	local edgeSize = size * 0.5

	for _, delta in ipairs(self.Deltas) do
		local _grid = self:GetGrid(coords + vector3(delta.x, delta.y, 0.0) * edgeSize, size, true)
		if not cache[_grid] then
			cache[_grid] = true
			grids[#grids + 1] = _grid
		end
	end

	return grids
end

File: resources/[core]/admin/node/grids.js
const Cfx = require("fivem-js")

module.exports = {
	Size: new Cfx.Vector3(8992, 12992, 0),
	Offset: new Cfx.Vector3(4480, 4480, 0),
	PlayerGridSize: 4,

	Deltas: [
		new Cfx.Vector3(-1.0, -1.0, 0),
		new Cfx.Vector3(-1.0, 0.0, 0),
		new Cfx.Vector3(-1.0, 1.0, 0),
		new Cfx.Vector3(0.0, -1.0, 0),
		new Cfx.Vector3(0.0, 1.0, 0),
		new Cfx.Vector3(1.0, -1.0, 0),
		new Cfx.Vector3(1.0, 0.0, 0),
		new Cfx.Vector3(1.0, 1.0, 0),
	],

	Sizes: [
		16,
		32,
		64,
		128,
	],

	GetWidth(size) {
		return Math.ceil(this.Size.x / size)
	},

	GetGrid(coords, size, keepSize) {
		coords = new Cfx.Vector3(Math.min(Math.max(coords.x + this.Offset.x, 0), this.Size.x), Math.min(Math.max(coords.y + this.Offset.y, 0), this.Size.y), coords.z)
		if (!keepSize) {
			size = this.Sizes[size]
		}

		var width = this.GetWidth(size)
		return Math.floor(coords.x / size) + Math.floor(coords.y / size) * width
	},

	GetCoords(grid, size, keepSize) {
		if (!keepSize) {
			size = this.Sizes[size]
		}

		var width = this.GetWidth(size)
		return new Cfx.Vector3(Math.floor(grid % width) * size - this.Offset.x + size / 2.0, Math.floor(grid / width) * size - this.Offset.y + size / 2.0, 0)
	},

	GetNearbyGrids(coords, size, keepSize) {
		if (!size) return {}
		if (!keepSize) {
			size = this.Sizes[size]
		}

		var width = this.GetWidth(size)
		if (typeof coords == "new Cfx.Vector3") {
			coords = this.GetGrid(coords, size, true)
		}

		var grid = coords

		return [
			grid,
			grid - 1,
			grid + 1,
			grid - width,
			grid - width - 1,
			grid - width + 1,
			grid + width,
			grid + width - 1,
			grid + width + 1
		]
	},

	GetImmediateGrids(coords, size, keepSize) {
		if (!size) return {}
		if (!keepSize) {
			size = this.Sizes[size]
		}

		if (typeof coords == "vector4") {
			coords = new Cfx.Vector3(coords.x, coords.y, coords.z)
		}

		var grid = this.GetGrid(coords, size, true)
		var cache = { [grid]: true }
		var grids = [ grid ]
		var edgeSize = size * 0.5

		this.Deltas.forEach(delta => {
			var _grid = this.GetGrid(coords + new Cfx.Vector3(delta.x, delta.y, 0.0) * edgeSize, size, true)
			if (!cache[_grid]) {
				cache[_grid] = true
				grids[grids.length + 1] = _grid
			}
		})

		return grids
	},
}